File: 09.math/solve_py/leetcode_69.py
def solve(x: int):
    """
    x = 4 -> 2
    x = 8 -> 2

    x = root(a)
    y <= root(a) < y+1
    y ** 2 <= a < (y+1) ** 2

    y 가 정답
    y = 0 일 때, a = 0
    y = 1 일 때, a = 1, 2, 3
    y = 2 일 때, a = 4, 5, 6, 7, 8
    y = 3 일 때, a = 9, 10, 11, 12, 13, 14, 15

    0 <= x <= 2^31 - 1
    0 <= root(x) <= root(2^31 - 1)

    모든 제곱수를 어떻게 찾지?
    가장 작은 정수부터 하나씩 제곱을 함 -> 그 값이 < x
    """
    y = 0
    sqrt = y * y

    while sqrt <= x:
        y += 1
        sqrt = y * y

    return y - 1

File: 09.math/solve_cpp/leetcode_69.cpp
/**
 * 개선 포인트 ? binary search 
 * sqrt = y * y 입장에서 0 과 x 사이이다. 
 * mid * mid < x -> left = mid
 * mid * mid > x -> right = mid
 * mid * mid == x -> return mid
 */
#include <iostream>

using namespace std;

class Solution {
    public:
        int sqrt(int x) {
            int left = 0;
            int right = x;

            while (left <= right) {
                int mid = (left + right) / 2;
                // long long square = (long long) mid * mid;
                int root = x;
                if (mid > 0) {
                    root /= mid;
                }

                if (mid < root) {
                    left = mid + 1;
                } else if (mid > root) {
                    right = mid - 1;
                } else {
                    return mid;
                }

                /**
                 * 아래 코드가 더 직관적이긴 함. 
                 * mid * mid 로 작성하면 integer overflow 발생 가능함.
                 * long long square = (long long) mid * mid;
                
                    if (square < x) {
                        left = mid + 1;
                    } else if (square > x) {
                        right = mid - 1;
                    } else {
                        return mid;
                    }
                 */
                
            }

            return right;
        }
};

int main() {
    Solution solution;

    for (int i = 0; i <= 30; i++) {
        printf("x = %d : %d\n", i, solution.sqrt(i));
    }

    return 0;
}